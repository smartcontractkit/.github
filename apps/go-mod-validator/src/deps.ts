import { execSync } from "child_process";
import { dirname, sep } from "path";
import { readFileSync } from "fs";
import * as core from "@actions/core";
import * as glob from "@actions/glob";
import { isPseudoVersion, pseudoVersionRev } from "./pseudo-version";

/**
 * Taken from `go help list` for the command `go list -json -m all`
 *  @link https://pkg.go.dev/cmd/go#hdr-List_packages_or_modules
 *  type Module struct {
 *        Path       string        // module path
 *        Query      string        // version query corresponding to this version
 *        Version    string        // module version
 *        Versions   []string      // available module versions
 *        Replace    *Module       // replaced by this module
 *        Time       *time.Time    // time version was created
 *        Update     *Module       // available update (with -u)
 *        Main       bool          // is this the main module?
 *        Indirect   bool          // module is only indirectly needed by main module
 *        Dir        string        // directory holding local copy of files, if any
 *        GoMod      string        // path to go.mod file describing module, if any
 *        GoVersion  string        // go version used in module
 *        Retracted  []string      // retraction information, if any (with -retracted or -u)
 *        Deprecated string        // deprecation message, if any (with -u)
 *        Error      *ModuleError  // error loading module
 *        Origin     any           // provenance of module
 *        Reuse      bool          // reuse of old module info is safe
 *    }
 *
 *    type ModuleError struct {
 *        Err string // the error itself
 *    }
 *
 */
interface GoMod {
  Path: string;
  Query: string;
  Version: string;
  Versions: string[];
  Replace?: GoMod;
  Time: Date | null;
  Update?: GoMod;
  Main: boolean;
  Indirect: boolean;
  Dir: string;
  GoMod: string;
  GoVersion: string;
  Retracted: string[];
  Deprecated: string;
  Error: ModuleError | null;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Origin: any;
  Reuse: boolean;
}

interface ModuleError {
  Err: string;
}

export interface BaseGoModule {
  /**
   * The name of the module is the path + version of the module,
   * along with a postfix of "// indirect" indicating if the module is indirect.
   */
  name: string;
  /**
   * Module path
   */
  path: string;
  /**
   * Module version
   */
  version: string;

  /**
   * The repo owner of the dependency.
   */
  owner: string;

  /**
   * The repo name of the dependency.
   */
  repo: string;

  /**
   * The path to the go.mod file describing the module.
   *
   * Not to be confused with "path", which is the module path, like github.com/smartcontractkit/go-proxy.
   * @example /path/to/go.mod
   */
  goModFilePath: string;
}

export interface GoModuleWithCommitSha extends BaseGoModule {
  commitSha: string;
}

export interface GoModuleWithTag extends BaseGoModule {
  tag: string;
}

export type GoModule = GoModuleWithCommitSha | GoModuleWithTag | BaseGoModule;

function parseGoModListOutput(jsonStr: string): GoMod[] {
  /*
   * {
   *   "Path": "github.com/smartcontractkit/libocr",
   *   "Version": "v0.0.0-20240419185742-fd3cab206b2c"
   * }
   * {
   *   "Path": "github.com/smartcontractkit/go-proxy",
   *   "Version": "v0.1.0"
   * }
   */

  // store parsed objects
  const goModules: GoMod[] = [];

  // keep concatenating lines until the object is complete
  let objStr = "";

  // keep track of the depth of the object
  let objDepth = 0;

  // Process
  for (const l of jsonStr.split("\n")) {
    if (l.includes("{")) {
      objDepth++;
    }
    if (l.includes("}")) {
      objDepth--;
    }

    objStr += l;
    if (objDepth === 0 && objStr) {
      goModules.push(JSON.parse(objStr));
      objStr = "";
    }
  }

  return goModules;
}

/**
 * Recursively gets the paths of all go.mod files in the specified directory.
 *
 * @param goModDir The directory to search for go.mod files.
 *
 * @throws If no go.mod files are found or an error occurs.
 */
export async function getAllGoModsWithin(goModDir: string): Promise<string[]> {
  let files: string[] = [];
  const pattern = `${goModDir}/**/go.mod`;
  try {
    const globber = await glob.create(pattern);
    files = await globber.glob();
    files = files.map((f) => f.replace(`${process.cwd()}${sep}`, `.${sep}`));
  } catch (error) {
    throw new Error(`failed to get go.mod files: ${error}`);
  }

  if (files.length == 0) {
    throw new Error(
      `no go.mod files found with pattern ${pattern} and cwd ${process.cwd()}`,
    );
  }
  return files;
}

/**
 * Gets all the dependencies of the Go modules in the specified directory.
 *
 * We recursively collect all go.mod files within a directory, then return the flattened list of all dependencies.
 * There may be duplicates in the list, but we don't care about that.
 *
 * Dependencies are filtered by depPrefix.
 *
 * @param rootDir The directory containing the Go modules.
 *
 */
export async function getDeps(
  rootDir: string,
  depPrefix: string,
): Promise<GoModule[]> {
  const modFilePaths = await getAllGoModsWithin(rootDir);

  const deps = modFilePaths.flatMap((goModFilePath) => {
    core.info(`Finding dependencies in ${goModFilePath}`);
    try {
      const dir = dirname(goModFilePath);
      const output = execSync("go list -json -m all", {
        encoding: "utf-8",
        cwd: dir,
      });
      core.debug(`Raw output: ${output}`);

      const parsedDeps = parseGoModListOutput(output);
      core.debug(`Parsed dependencies: ${JSON.stringify(parsedDeps)}`);

      const mappedDeps = goModsToGoModules(
        goModFilePath,
        parsedDeps,
        depPrefix,
      );
      core.debug(`Mapped dependencies: ${JSON.stringify(mappedDeps)}`);

      return mappedDeps;
    } catch (error) {
      throw Error(
        `failed to get go.mod dependencies from file: ${goModFilePath}, err: ${error}`,
      );
    }
  });

  return deps;
}

/**
 * Creates a function that finds the line number of a dependency path in a go.mod file.
 */
export function lineForDependencyPathFinder() {
  const cache: {
    // goModPath -> [go mod file lines]
    [goModPath: string]: string[];
  } = {};

  /**
   * Given a go.mod file path and a dependency path, returns the line number of the dependency path in the go.mod file.
   *
   * We assume that you cannot have two modules of the same name in the go.mod file.
   * Technically, you can, but you need to make use of the replace directive along with two copies of the same module locally in order to do so.
   * This seems to be a rare case, so we are not handling it.
   *
   * @param goModPath The path to the go.mod file. E.g. /path/to/go.mod
   * @param depPath The dependency path to search for in the go.mod file. E.g. github.com/smartcontractkit/go-proxy
   *
   * @throws If the dependency path is found more than once in the go.mod file.
   */
  return function getDepPath({
    goModFilePath,
    path,
    name,
  }: BaseGoModule): number {
    core.debug(
      `Finding line number for ${path} in ${goModFilePath}. (${name})`,
    );
    if (!cache[goModFilePath]) {
      cache[goModFilePath] = readFileSync(goModFilePath, "utf-8")
        .split("\n")
        .map((l) => l.trim());
    }

    let line = -1;
    for (let i = 0; i < cache[goModFilePath].length; i++) {
      // HACK: We add a space after the depPath to avoid matching substrings.
      if (cache[goModFilePath][i].includes(path + " ")) {
        if (line !== -1) {
          core.warning(
            `Duplicate dependency path found: ${path} in ${goModFilePath} (line ${i + 1}). Previously found on line ${line}. Annotations may be faulty. Skipping.`,
          );
          continue;
        }
        line = i + 1;
      }
    }

    if (line === -1) {
      throw new Error(`dependency path not found: ${path}`);
    }

    return line;
  };
}

/**
 * Takes the parsed output of `go list -json -m all` and converts it to a list of GoModule objects.
 *
 * Performs the following transformations:
 * - Filters out the main module and non-org dependencies.
 * - Maps the GoMod object to a GoModule object.
 * @param goMods
 */
export function goModsToGoModules(
  goModFilePath: string,
  goMods: GoMod[],
  depPrefix: string,
): GoModule[] {
  const goModules = goMods
    // Replace the module with the Replace field if it exists
    .map((d) => d.Replace || d)
    // `go list -m -json all` also lists the main package, avoid parsing it.
    // and only validate dependencies belonging to our org
    .filter((d) => !d.Main && d.Path.startsWith(depPrefix))
    .map((d: GoMod): GoModule => {
      // repo format github.com/smartcontractkit/chainlink
      const [, owner, repo] = d.Path.split("/");
      const baseModule: BaseGoModule = {
        owner,
        repo,
        goModFilePath,
        name: `${d.Path}@${d.Version}${d.Indirect ? " // indirect" : ""}`,
        path: d.Path,
        version: d.Version,
      };

      const versionType = getVersionType(d.Version);

      if (!versionType.tag && !versionType.commitSha) {
        core.warning(
          `invalid version string: ${d.Version} for module: ${d.Path}`,
        );
        return baseModule;
      } else if (versionType.commitSha) {
        return {
          ...baseModule,
          commitSha: versionType.commitSha,
        };
      } else {
        return {
          ...baseModule,
          tag: versionType.tag,
        };
      }
    });

  return goModules;
}

/**
 * Extracts the type of a given version string.
 *
 * @param versionString - The version string to extract the type from.
 *
 * A pseudo version is a version string that is generated by Go when a module is built from a commit that is not tagged. Thus, they are generated from commit data.
 * See: https://go.dev/ref/mod#pseudo-versions for more details.
 *
 * A regular version is a version string that is a tag.
 *
 */
export function getVersionType(versionString: string) {
  // matches real versions like v0.1.0
  const versionRegex = /^(v\d+\.\d+\.\d+)$/;

  if (isPseudoVersion(versionString)) {
    return {
      commitSha: pseudoVersionRev(versionString),
      tag: undefined,
    };
  }
  const versionMatch = versionString.match(versionRegex);
  if (versionMatch) {
    return {
      commitSha: undefined,
      tag: versionMatch?.[1],
    };
  }

  return {
    commitSha: undefined,
    tag: undefined,
  };
}
