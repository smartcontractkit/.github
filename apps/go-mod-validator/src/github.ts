import { Octokit } from "octokit";

async function getDefaultBranch(
  owner: string,
  repo: string,
  octokitClient: Octokit,
): Promise<string> {
  const repoObject = await octokitClient.request(
    `GET /repos/${owner}/${repo}/`,
  );
  return repoObject.data.default_branch;
}

async function findCommitInDefaultBranch(
  owner: string,
  repo: string,
  defaultBranch: string,
  commitSha: string,
  octokitClient: Octokit,
) {
  const compareResult = await octokitClient.request(
    `GET /repos/${owner}/${repo}/compare/${defaultBranch}...${commitSha}`,
  );
  return (
    compareResult.data.status === "identical" ||
    compareResult.data.status === "behind"
  );
}

async function findTagInDefaultBranch(
  owner: string,
  repo: string,
  defaultBranch: string,
  tag: string,
  octokitClient: Octokit,
) {
  const repoTags = await octokitClient.request(
    `GET /repos/${owner}/${repo}/tags`,
  );
  for (const repoTag of repoTags.data) {
    if (repoTag.name != tag) {
      continue;
    }

    // check commit is on the default branch
    return await findCommitInDefaultBranch(
      owner,
      repo,
      defaultBranch,
      repoTag.commit.sha,
      octokitClient,
    );
  }

  return false;
}

async function getVersionType(versionString: string) {
  // matches pseudo versins like v0.0.5-0.20220116011046-fa5810519dcb
  const pseudoVersionRegex = /-([\d.]*)-([a-f0-9]{12})(?:\+[\w.-]+)?$/;
  // matches real versions like v0.1.0
  const versionRegex = /^(v\d+\.\d+\.\d+)$/;

  const pseudoVersionMatch = versionString.match(pseudoVersionRegex);
  const versionMatch = versionString.match(versionRegex);

  if (pseudoVersionMatch) {
    return {
      commitSha: pseudoVersionMatch[2],
      tag: null,
    };
  }
  if (versionMatch) {
    return {
      commitSha: null,
      tag: versionMatch[1],
    };
  }
}

export async function validateDependency(
  path: string,
  version: string,
  octokitClient: Octokit,
) {
  // repo format smartcontractkit/chainlink
  const repoPathSplit = path.split("/");
  const owner = repoPathSplit[1];
  const repo = repoPathSplit[2];

  const defaultBranch = await getDefaultBranch(owner, repo, octokitClient);

  const result = await getVersionType(version);
  if (result?.commitSha) {
    return await findCommitInDefaultBranch(
      owner,
      repo,
      defaultBranch,
      result.commitSha,
      octokitClient,
    );
  }
  if (result?.tag) {
    return await findTagInDefaultBranch(
      owner,
      repo,
      defaultBranch,
      result.tag,
      octokitClient,
    );
  }

  return false;
}
