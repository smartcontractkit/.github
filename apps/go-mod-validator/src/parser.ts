const smartContractKitPrefix = "github.com/smartcontractkit/";

function parseGoModLine(line: string) {
  // Trim any extra whitespace
  line = line.trim();

  // Go modules have psuedo versions and real versions
  // 1. psuedo version: v1.1.1-0.20190123174540-a2c9a5303de7
  // 2. real versions: v1.0.2
  // also, there can be pre-release versions
  // 1. v1.1.1-0.20190123174540-a2c9a5303de7+incompatible
  // 2. v1.0.2-beta
  // more details on - https://go.dev/doc/modules/version-numbers
  const moduleRegex = new RegExp(
    /^([\w.-]+(?:\/[\w.-]+)*) (v\d+\.\d+\.\d+(?:-[\w\d.-]+)?(?:\+[\w\d.-]+)?)$/,
  );
  const replaceRegex = new RegExp(
    moduleRegex.source.slice(0, moduleRegex.source.length - 1) +
      " => " +
      moduleRegex.source.slice(1),
  );
  const moduleMatch = line.match(moduleRegex);
  const replaceMatch = line.match(replaceRegex);

  if (moduleMatch) {
    const [, module, version] = moduleMatch;
    return {
      module,
      version,
    };
  }

  if (replaceMatch) {
    const [, , , module, version] = replaceMatch;
    return {
      module,
      version,
    };
  }

  return null;
}

// extract all final dependencies
function parseGoModLines(lines: string[]) {
  const dependencies: { module: string; version: string }[] = [];

  // exclude the first line which is module information
  lines = lines.slice(1);

  // parse each line to find out the final dependency
  lines.forEach((line) => {
    if (line.length == 0) {
      return;
    }

    const result = parseGoModLine(line);
    if (result) {
      console.debug(
        `Parsed Module: ${result.module}, Version: ${result.version}`,
      );
      dependencies.push(result);
    } else {
      console.error(`Could not parse line: ${line}`);
      throw new Error(`Could not parse line: ${line}`);
    }
  });

  return dependencies;
}

// extract only smartcontractkit dependencies
export function getSmartcontractkitDependencies(goListOutputContent: string) {
  // Split the content into lines
  const lines = goListOutputContent.split("\n");

  const dependencies: { module: string; version: string }[] = [];
  parseGoModLines(lines).forEach((dependency) => {
    if (dependency.module.startsWith(smartContractKitPrefix)) {
      dependencies.push(dependency);
    }
  });

  return dependencies;
}
