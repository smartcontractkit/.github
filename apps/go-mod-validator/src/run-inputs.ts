import * as core from "@actions/core";

export const CL_LOCAL_DEBUG = process.env.CL_LOCAL_DEBUG === "true";

export interface RunInputs {
  githubToken: string;
  goModDir: string;
  depPrefix: string;
}

export function getInputs(): RunInputs {
  core.info("Getting inputs for run.");

  const inputs: RunInputs = {
    githubToken: getRunInputString("githubToken"),
    goModDir: getRunInputString("goModDir"),
    depPrefix: getRunInputString("depPrefix"),
  };

  core.info(`Inputs: ${JSON.stringify(inputs)}`);
  return inputs;
}

interface RunInputConfiguration {
  /**
   * The parameter name as defined in the action.yml file.
   */
  parameter: string;
  /**
   * The local environment variable name to use when debugging locally.
   * This is typically the parameter name with dashes replaced by underscores.
   * For example, "directory" becomes "DIRECTORY". foo-bar becomes FOO_BAR.
   */
  localParameter: string;
}

const runInputsConfiguration: {
  [K in keyof RunInputs]: RunInputConfiguration;
} = {
  githubToken: {
    parameter: "github-token",
    localParameter: "GITHUB_TOKEN",
  },
  goModDir: {
    parameter: "go-mod-dir",
    localParameter: "GO_MOD_DIR",
  },
  depPrefix: {
    parameter: "dep-prefix",
    localParameter: "DEP_PREFIX",
  },
};

function getRunInputString(input: keyof RunInputs) {
  const inputKey = getInputKey(input);
  return core.getInput(inputKey, {
    required: true,
  });
}

function getInputKey(input: keyof RunInputs) {
  const config = runInputsConfiguration[input];
  if (!config) {
    // this should never happen due to type safety
    throw new Error(`No configuration found for input: ${input}`);
  }

  // Use local debug input key if local debugging is enabled
  const inputKey = CL_LOCAL_DEBUG ? config.localParameter : config.parameter;
  return inputKey;
}
