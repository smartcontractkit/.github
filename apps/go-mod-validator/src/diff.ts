import * as core from "@actions/core";
import { GetResponseTypeFromEndpointMethod } from "@octokit/types";
import { Octokit } from "./github";
type CompareResponse = GetResponseTypeFromEndpointMethod<
  Octokit["rest"]["repos"]["compareCommitsWithBasehead"]
>;
export type GithubFiles = CompareResponse["data"]["files"];

export async function getChangedGoModFiles(
  gh: Octokit,
  base: string,
  head: string,
  owner: string,
  repo: string,
  depPrefix: string,
): Promise<ParsedFile[]> {
  console.log({
    base,
    head,
    owner,
    repo
  })

  const files = await getComparison(gh, owner, repo, base, head);
  const relevantFiles = filterForRelevantChanges(files, );
  return parseAllAdditions(relevantFiles, depPrefix);
}

 function filterForRelevantChanges(files: GithubFiles): GithubFiles {
  return files?.filter(({ filename }) => filename.endsWith("go.mod"));
}


 async function getComparison(
  octokit: Octokit,
  owner: string,
  repo: string,
  base: string,
  head: string,
): Promise<GithubFiles> {
  core.debug(`Comparing ${owner}/${repo} commits ${base}...${head}`);

  const diff = await octokit.rest.repos.compareCommitsWithBasehead({
    owner,
    repo,
    basehead: `${base}...${head}`,
    // <before>...<after> or <earlier>...<later>
  });

  return diff.data.files;
}

// Taken from https://github.com/smartcontractkit/.github/blob/dc8b1a0b478151119d86ac1bf121ea7eb3c1c88c/actions/gha-workflow-validator/src/utils.ts#L59C1-L97C2
export interface ParsedFile {
  filename: string;
  sha: string;
  addedLines: FileAddition[];
}

export interface FileAddition {
  lineNumber: number;
  content: string;
}

 function parseAllAdditions(files: GithubFiles, depPrefix: string): ParsedFile[] {
  if (!files) return [];

  return files?.map((entry) => {
    const { filename, sha, patch } = entry;
    const addedLines = patch ? parsePatchAdditions(patch, depPrefix) : [];
    return { filename, sha, addedLines };
  });
}

 function parsePatchAdditions(patch: string, depPrefix: string): FileAddition[] {
  const lineChanges = patch?.split("\n") || [];

  const additions: FileAddition[] = [];

  let currentLineInFile = 0;
  for (const line of lineChanges) {
    if (line.startsWith("@@")) {
      // @@ denotes a git hunk header
      // https://mattscodecave.com/posts/howto-reading-git-diffs-and-staging-hunks.html
      // example line: @@ -16,10 +16,10 @@ jobs:
      //   - "-16,10": "-" denotes source file, "16,10" means the hunk starts at line 16 and output contains 10 lines
      //   - "+16,10": "+" denotes destination file, "16,10" means the same as above
      const [, , destination] = line.split(" ");

      if (!destination.startsWith("+")) {
        throw new Error("Invalid git hunk format");
      }

      const [destinationLine] = destination.substring(1).split(",");
      currentLineInFile = parseInt(destinationLine, 10);
      continue;
    } else if (line.startsWith("+")) {
      const currentLine = line.substring(1);
      if (isGoModule(currentLine, depPrefix)) {
        additions.push({
          content: currentLine,
          lineNumber: currentLineInFile,
        });
      }
    } else if (line.startsWith("-")) {
      // ignore deletions
      continue;
    }
    currentLineInFile++;
  }

  return additions;
}

function isGoModule(line: string, depPrefix: string): boolean {
  // Quick and dirty, we just search for github.com and include it in the results
  return line.includes(depPrefix);
}
