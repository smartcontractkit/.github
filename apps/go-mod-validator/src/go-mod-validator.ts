import { isGoModReferencingDefaultBranch } from "./github";
import { getDeps } from "./deps";
import { FIXING_ERRORS } from "./strings";
import * as github from "@actions/github";
import * as core from "@actions/core";
import { throttling } from "@octokit/plugin-throttling";

function getContext() {
  const goModDir = core.getInput("go-mod-dir", { required: true });
  const githubToken = core.getInput("github-token", { required: true });
  const depPrefix = core.getInput("dep-prefix", { required: true });
  const gh = github.getOctokit(
    githubToken,
    {
      throttle: {
        onRateLimit: (retryAfter, options, octokit, retryCount) => {
          octokit.log.warn(
            `Request quota exhausted for request ${options.method} ${options.url}`,
          );

          if (retryCount < 1) {
            // only retries once
            octokit.log.info(`Retrying after ${retryAfter} seconds!`);
            return true;
          }
        },
        onSecondaryRateLimit: (retryAfter, options, octokit) => {
          // does not retry, only logs a warning
          octokit.log.warn(
            `SecondaryRateLimit detected for request ${options.method} ${options.url}`,
          );
        },
      },
    },
    throttling,
  );

  return { goModDir, gh, depPrefix };
}

export async function run(): Promise<string> {
  const { goModDir, gh, depPrefix } = getContext();

  const depsToValidate = await getDeps(goModDir, depPrefix);

  // <dependency-name, error-string>
  const errs: Map<string, string> = new Map();

  const validating = depsToValidate.map(async (d) => {
    const isValid = await isGoModReferencingDefaultBranch(d, gh);

    if (!isValid) {
      errs.set(d.name, "dependency not on default branch");
    }
  });

  await Promise.all(validating);
  if (errs.size > 0) {
    errs.forEach((depName, errMsg) =>
      core.error(`validation failed for: ${errMsg}, err: ${depName}`),
    );

    core.summary.addRaw(FIXING_ERRORS, true);
    const summary = core.summary.stringify();
    await core.summary.write();

    core.setFailed(`validation failed for ${errs.size} dependencies`);
    return summary;
  } else {
    const msg = "validation successful for all go.mod dependencies";
    core.info(msg);
    return msg;
  }
}
