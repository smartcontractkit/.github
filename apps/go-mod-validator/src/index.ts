import { validateDependency } from "./github";
import { getDependenciesMap } from "./deps";

const smartContractKitPrefix = "github.com/smartcontractkit";

async function run() {
  // Read required env vars
  const githubToken = process.env.GITHUB_TOKEN || "";
  if (githubToken == "") {
    console.error("no GITHUB_TOKEN env variable found");
    process.exit(1);
  }

  // get dependencies from go.mod file
  const dependenciesMap = getDependenciesMap();

  // Verify each of the dependencies
  let validationErr = null;
  for (let [file, dependencies] of dependenciesMap.entries()) {
    console.info(`\nvalidating dependencies for ${file}`);

    for (let dependency of dependencies) {
      // handle replace redirectives
      if (dependency.Replace != undefined) {
        dependency = dependency.Replace;
      }

      // `go list -m -json all` also lists the main pacakge, avoid parsing it.
      // and only validate dependencies belonging to our org
      if (
        dependency.Version == undefined ||
        !dependency.Path.startsWith(smartContractKitPrefix)
      ) {
        continue;
      }

      try {
        if (
          await validateDependency(
            dependency.Path,
            dependency.Version,
            githubToken,
          )
        ) {
          console.info(
            `${dependency.Path}@${dependency.Version} is found in the default branch`,
          );
        } else {
          console.error(
            `${dependency.Path}@${dependency.Version} not found in the default branch`,
          );
          validationErr = new Error(
            `${dependency.Path}@${dependency.Version} not found in the default branch`,
          );
        }
      } catch (err) {
        console.error(
          `failed to verify dependency: ${dependency.Path}@${dependency.Version}, with err: ${err}`,
        );
        validationErr = err;
      }
    }
  }

  if (validationErr != null) {
    process.exit(1);
  }
}

run();
