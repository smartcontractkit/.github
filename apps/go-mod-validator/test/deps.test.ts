import { execSync } from "child_process";
import { getDeps, getVersionType } from "../src/deps";
import { describe, expect, it, vi, MockedObject } from "vitest";
import * as glob from "@actions/glob";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
vi.mock("@actions/core", async (importOriginal: any) => ({
  ...(await importOriginal(typeof import("@actions/core"))),
  setFailed: (msg: string) => {
    console.log(`setFailed (stub): ${msg}`);
  },
  error: (msg: string) => {
    console.log(`error (stub): ${msg}`);
  },
  warning: (msg: string) => {
    console.log(`warn (stub): ${msg}`);
  },
  info: (msg: string) => {
    console.log(`info (stub): ${msg}`);
  },
  debug: () => {
    // noop
  },
}));

const mockedExecSync = vi.mocked(execSync);
vi.mock("child_process");
const mockedGlob = vi.mocked(glob);
vi.mock("@actions/glob");
type MockedGlob = MockedObject<Awaited<ReturnType<typeof glob.create>>>;

describe("getDependenciesMap", () => {
  it("should return a map of <go.mod files: dependencies in json>", async () => {
    const paths = ["/path/to/first/go.mod", "/path/to/second/go.mod"];
    const goList1 =
      '{"Path": "github.com/smartcontractkit/go-plugin", "Version": "v0.0.0-20240208201424-b3b91517de16"}';
    const goList2 =
      '{"Path": "github.com/smartcontractkit/grpc-proxy", "Version": "v0.0.0-20230731113816-f1be6620749f"}';
    mockedGlob.create.mockResolvedValueOnce({
      glob: vi.fn().mockResolvedValue(paths),
    } as MockedGlob);

    mockedExecSync.mockImplementationOnce(() => goList1);
    mockedExecSync.mockImplementationOnce(() => goList2);

    const result = await getDeps("", "github.com/smartcontractkit");
    expect(result).toMatchInlineSnapshot(`
      [
        {
          "commitSha": "b3b91517de16",
          "goModFilePath": "/path/to/first/go.mod",
          "name": "github.com/smartcontractkit/go-plugin@v0.0.0-20240208201424-b3b91517de16",
          "owner": "smartcontractkit",
          "path": "github.com/smartcontractkit/go-plugin",
          "repo": "go-plugin",
          "version": "v0.0.0-20240208201424-b3b91517de16",
        },
        {
          "commitSha": "f1be6620749f",
          "goModFilePath": "/path/to/second/go.mod",
          "name": "github.com/smartcontractkit/grpc-proxy@v0.0.0-20230731113816-f1be6620749f",
          "owner": "smartcontractkit",
          "path": "github.com/smartcontractkit/grpc-proxy",
          "repo": "grpc-proxy",
          "version": "v0.0.0-20230731113816-f1be6620749f",
        },
      ]
    `);
  });

  it("should handle no go.mod files found", async () => {
    const paths: string[] = [];

    mockedGlob.create.mockResolvedValue({
      glob: vi.fn().mockResolvedValue(paths),
    } as MockedGlob);

    await expect(getDeps("", "")).rejects.toThrow("no go.mod files found");
  });

  it("should handle glob search failure", async () => {
    mockedGlob.create.mockRejectedValue(new Error("Glob error"));
    await expect(getDeps("", "")).rejects.toThrow("Glob error");
  });
});

// Mock data for tests
// https://cs.opensource.google/go/x/mod/+/refs/tags/v0.20.0:module/pseudo_test.go
// https://cs.opensource.google/go/x/mod/+/refs/tags/v0.20.0:module/pseudo.go;l=164
const pseudoTests = [
  { major: "", older: "", version: "v0.0.0-20060102150405-hash" },
  { major: "v0", older: "", version: "v0.0.0-20060102150405-hash" },
  { major: "v1", older: "", version: "v1.0.0-20060102150405-hash" },
  { major: "v2", older: "", version: "v2.0.0-20060102150405-hash" },
  { major: "unused", older: "v0.0.0", version: "v0.0.1-0.20060102150405-hash" },
  { major: "unused", older: "v1.2.3", version: "v1.2.4-0.20060102150405-hash" },
  {
    major: "unused",
    older: "v1.2.99999999999999999",
    version: "v1.2.100000000000000000-0.20060102150405-hash",
  },
  {
    major: "unused",
    older: "v1.2.3-pre",
    version: "v1.2.3-pre.0.20060102150405-hash",
  },
  {
    major: "unused",
    older: "v1.3.0-pre",
    version: "v1.3.0-pre.0.20060102150405-hash",
  },
  {
    major: "unused",
    older: "v0.0.0--",
    version: "v0.0.0--.0.20060102150405-hash",
  },
  {
    major: "unused",
    older: "v1.0.0+metadata",
    version: "v1.0.1-0.20060102150405-hash+metadata",
  },
  {
    major: "unused",
    older: "v2.0.0+incompatible",
    version: "v2.0.1-0.20060102150405-hash+incompatible",
  },
  {
    major: "unused",
    older: "v2.3.0-pre+incompatible",
    version: "v2.3.0-pre.0.20060102150405-hash+incompatible",
  },
];

describe("getVersionType", () => {
  for (const { version } of pseudoTests) {
    it(`should return the correct version type for ${version} `, () => {
      const verType = getVersionType(version);
      expect(verType).toEqual({ commitSha: "hash", tag: undefined });
    });
  }

  it("should parse out the git sha correctly", () => {
    let version = "v0.2.2-0.20240808143317-6b16fc28887d";
    let verType = getVersionType(version);
    expect(verType).toEqual({ commitSha: "6b16fc28887d", tag: undefined });

    version = "v0.0.1-beta-test.0.20240709043547-03612098f799";
    verType = getVersionType(version);
    expect(verType).toEqual({ commitSha: "03612098f799", tag: undefined });
  });

  it("should parse out the tag correctly", () => {
    const version = "v0.0.1";
    const verType = getVersionType(version);
    expect(verType).toEqual({ commitSha: undefined, tag: "v0.0.1" });
  });
});
