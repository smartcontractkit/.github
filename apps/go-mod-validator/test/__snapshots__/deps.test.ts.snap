// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getVersionType > should return the correct version type for v0.0.0--.0.20060102150405-hash  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v0.0.0-20060102150405-hash  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v0.0.0-20060102150405-hash  2`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v0.0.1-0.20060102150405-hash  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v0.0.5-0.20220116011046-fa5810519dcb  1`] = `
{
  "commitSha": "fa5810519dcb",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v1.0.0-20060102150405-hash  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v1.0.1-0.20060102150405-hash+metadata  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v1.2.3-pre.0.20060102150405-hash  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v1.2.4-0.20060102150405-hash  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v1.2.100000000000000000-0.20060102150405-hash  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v1.3.0-pre.0.20060102150405-hash  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v2.0.0-20060102150405-hash  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v2.0.1-0.20060102150405-hash+incompatible  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type for v2.3.0-pre.0.20060102150405-hash+incompatible  1`] = `
{
  "commitSha": "hash",
  "tag": undefined,
}
`;
